### PR TITLE
[Documentation] MLIR Plugins

### PR DESCRIPTION
**Pull request by erick-xanadu**
**Context:** MLIR Plugins need to be documented. In particular a small how to develop an MLIR plugin that is compatible with Catalyst and use it from within Python.

**Description of the Change:** Just documentation changes. 

**Benefits:** Users can immediately start working on building standalone plugins compatible with Catalyst.

**Possible Drawbacks:** 

[sc-73569]

### CODE REVIEW

**Review by dime10 — COMMENTED** [review]
Thanks @erick-xanadu, this is a nice write-up!

Only a few minors corrections. In terms of the general style, I would recommend to tone down the "handholdy" or "schooling" style a little bit (ideally we are targeting component engineers with this), but that could be personal preference so I'll leave that up to you.

File: doc/dev/plugins.rst
[the remark this answers is not in the document]
Comment (erick-xanadu) [reply doc/dev/plugins.rst:1]: The standalone dialect is an out of tree dialect. I had some issues using `find_package` to obtain Catalyst's configuration. @mlxd and I discussed this briefly and he has some ideas about changing Catalyst CMake packages to make it easier to do this. We agreed that for the time being, just adding the plugin into Catalyst is a sufficient good enough solution. But I agree that something like what we use with MHLO should be used here.

File: doc/dev/plugins.rst [hunk 1]
```diff
@@ -0,0 +1,357 @@
...
+      }
+    }
+
+But if we try to run it, using the same command as shown earlier 
+
+.. code-block::
+
+      quantum-opt --load-pass-plugin=/path/to/StandalonePlugin.so --pass-pipeline='builtin.module(standalone-switch-bar-to-foo) example.mlir'
+
+the compilation will fail with a message similar to:
```
[the remark this answers is not in the document]
Comment (erick-xanadu) [reply doc/dev/plugins.rst]: I think it is fine as it is.

**Comment by erick-xanadu** [comment]
> Only a few minors corrections. In terms of the general style, I would recommend to tone down the "handholdy" or "schooling" style a little bit (ideally we are targeting component engineers with this), but that could be personal preference so I'll leave that up to you.

I prefer to keep the style as easy as possible. We don't know who will be writing compilation passes. It could be anyone.

File: doc/dev/plugins.rst
[the remark this answers is not in the document]
Comment (dime10) [reply doc/dev/plugins.rst:1]: Got it 👍

**Comment by dime10** [comment]
> > Only a few minors corrections. In terms of the general style, I would recommend to tone down the "handholdy" or "schooling" style a little bit (ideally we are targeting component engineers with this), but that could be personal preference so I'll leave that up to you.
> 
> I prefer to keep the style as easy as possible. We don't know who will be writing compilation passes. It could be anyone.

I was thinking of phrases like `start your quantum compilation journey` which are just fluff (no effect on content or difficulty), but change the tone from technical document to tutorial for younger audiences.

From my side I don't have anything to add to the review, I'll leave it to Josh to read it over as well.

**Comment by erick-xanadu** [comment]
Sure I can delete that fluff. I'll leave it up for review and see if Josh has time to review. Otherwise, I'll change it a bit.

**Comment by erick-xanadu** [comment]
@dime10 removed some comments from the documentation.

File: doc/dev/plugins.rst [hunk 2]
```diff
@@ -0,0 +1,356 @@
...
+    +    // failure indicates that nothing was modified.
+         return failure();
+       }
+     };
+
+And then we can run ``make all`` again.
+The shared object of the standalone plugin should be available in ``mlir/build/lib/StandalonePlugin.so``.
+This shared object can be used with ``catalyst-cli`` and ``quantum-opt``.
+You can change the name of the pass, change the name of the shared object and make any changes you want.
+This was just an easy example to get started.
```
Comment (dime10) [inline doc/dev/plugins.rst:331]:
```suggestion
From here, you can change the name of the pass, change the name of the shared object, and implement more complex transformations.
```

**Review by dime10 — APPROVED** [review]
Thanks Erick, happy to approve :)

Some suggestions to further trim the fat if you're so inclined 👍

**Pull request merged**